### PR TITLE
feat: add subtask and dependency commands (#16)

### LINKED ISSUE #16: Phase 3: Subtasks & Dependencies
## Overview
Support task hierarchies and dependency relationships for complex workflows.

**Priority**: High  
**Timeline**: Week 5-6  
**Phase**: 3 of 6

## Features

- Create and manage subtasks
- Set task dependencies (blocked by / blocking)
- Convert existing tasks to subtasks
- Visualize task relationships

## Commands

```bash
# Subtask operations
asana task subtask list <parent-id>
asana task subtask create <parent-id> --name "Subtask" --notes "Details"
asana task subtask add <task-id> <parent-id>

# Dependency operations
asana task dependency add <task-id> <depends-on-id>
asana task dependency remove <task-id> <depends-on-id>
asana task dependency list <task-id>
asana task dependent list <task-id>  # Tasks that depend on this one
```

## Technical Approach

- Extend `src/commands/task.ts` with subcommands
- Use `TasksApi.setParentForTask()` for subtasks
- Use `TasksApi.addDependenciesForTask()` and `TasksApi.addDependentsForTask()`
- Handle API limitation: max 30 dependencies + dependents combined per task
- Add validation and clear error messages for limits
- Support recursive subtask listing

## Deliverables

- [ ] Implement subtask commands (list, create, add)
- [ ] Implement dependency commands (add, remove, list)
- [ ] Add validation for dependency limits
- [ ] Unit tests with edge case coverage
- [ ] E2E tests for complex task hierarchies
- [ ] Documentation with workflow examples
- [ ] Update CHANGELOG.md

## Testing Requirements

- Test subtask hierarchies (multiple levels)
- Test dependency cycles (should fail gracefully)
- Test dependency limit (30 combined)
- Test converting task to subtask
- Mock complex task relationships
- Test coverage >80%

## Success Criteria

- ✅ Subtask creation and management works
- ✅ Dependency management with proper limits
- ✅ Clear error messages for API limitations
- ✅ All tests passing
- ✅ Documentation with complex workflow examples

## Related Documentation

- [ROADMAP.md](https://github.com/pleaseai/asana/blob/main/dev-docs/ROADMAP.md)
- [Asana Tasks API](https://github.com/Asana/node-asana/blob/master/docs/TasksApi.md)

## Dependencies

- Phase 1: Task Update Operations (for task manipulation)
- Phase 2: Project Management (for project context)

## Blocks

None

### PR DESCRIPTION
**Pull request by amondnet**
## Summary

Implements **Phase 3: Subtasks & Dependencies** (#16) — task hierarchies and
blocked-by/blocking relationships for the Asana CLI.

## Commands added

```bash
# Subtasks
asana task subtask list <parent-id> [--recursive]
asana task subtask create <parent-id> --name "..." [--notes ...] [--assignee ...] [--due-on YYYY-MM-DD]
asana task subtask add <task-id> <parent-id>     # re-parent an existing task

# Dependencies (this task is blocked by another)
asana task dependency add <task-id> <depends-on-id>
asana task dependency remove <task-id> <depends-on-id>
asana task dependency list <task-id>

# Dependents (tasks blocked by this task)
asana task dependent list <task-id>
```

## Implementation notes

- New command modules `task-subtask.ts` and `task-dependency.ts`, wired into the
  `task` command, to keep each file focused.
- Extended the `asana-client` wrapper with `getSubtasks`, `createSubtask`,
  `setParent`, and dependency/dependent get/add/remove methods
  (`TasksApi.*ForTask`).
- `validateDependencyLimit` enforces the combined **30** dependencies + dependents
  cap before calling the API; `validateNoSelfDependency` rejects self-dependencies.
  Both are pure functions with full unit coverage.
- Added `getOutputFormat()` helper using `optsWithGlobals()` so the global
  `--format` flag resolves correctly for nested subcommands.
- Recursive subtask listing flattens the tree with a `depth` column.

## Testing

- `bun run test:unit` — **192 pass / 0 fail** (added validator, subtask, and
  dependency command tests; updated the command-ordering test).
- `bun run lint` — clean.
- E2E (`tests/e2e/subtasks-dependencies.test.ts`): covers multi-level
  hierarchies, dependency add/list/remove, self-dependency and circular
  dependency rejection. Requires real Asana credentials (`test:e2e:secure`).

## Notes / Decisions

- **CHANGELOG.md** is managed by release-please (generated from conventional
  commits on release), consistent with how Phase 1/2 were handled, so it is not
  hand-edited here.
- Deep cycle detection is delegated to the Asana API, which rejects cycles; the
  error handler surfaces a clear message. Local validation covers self-dependency
  and the combined limit.

Closes #16

### CODE REVIEW

File: src/commands/task-subtask.ts [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      name: subtask.name,
+      completed: subtask.completed ?? false,
+      depth,
+    })
+    const nested = await collectSubtasks(client, subtask.gid, depth + 1)
+    entries.push(...nested)
+  }
+
+  return entries
+}
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/commands/task-subtask.ts:55]: [claude] @gemini-code-assist 적용 완료 ✅ `collectSubtasks`에 `{ opt_fields: 'name,completed' }`를 추가해 completed 누락을 해결하고, 형제 서브트리를 `Promise.all`로 병렬 수집하도록 변경했으며 순환 데이터 방지용 `visited` Set도 도입했습니다. (2ee95fb)

File: src/commands/task-subtask.ts [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+          subtaskList = await collectSubtasks(client, parentGid, 0)
+        }
+        else {
+          const response = await client.tasks.getSubtasks(parentGid)
+          subtaskList = (response.data || []).map((item: any) => ({
+            gid: item.gid,
+            name: item.name,
+            completed: item.completed ?? false,
+            depth: 0,
+          }))
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/commands/task-subtask.ts]: [claude] @gemini-code-assist 적용 완료 ✅ 비재귀 `task subtask list` 경로도 `{ opt_fields: 'name,completed' }`를 전달하도록 수정해 completed 상태가 정상 출력됩니다. (2ee95fb)

File: src/commands/task-dependency.ts [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+
+        // Asana enforces a combined limit on dependencies + dependents.
+        const [dependencies, dependents] = await Promise.all([
+          client.tasks.getDependencies(taskGid, RELATIONSHIP_FIELDS),
+          client.tasks.getDependents(taskGid, RELATIONSHIP_FIELDS),
+        ])
+        const combined = (dependencies.data?.length || 0) + (dependents.data?.length || 0)
+        validateDependencyLimit(combined, 1)
+
+        await client.tasks.addDependencies(taskGid, [dependsOnGid])
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/commands/task-dependency.ts]: [claude] @gemini-code-assist 적용 완료 ✅ 추가 전에 기존 의존성 존재 여부를 확인하여, 이미 등록된 경우 한도 검증과 API 호출을 건너뛰도록 했습니다. 이로써 한도(30개)에 도달한 작업에서도 기존 의존성 재확인이 오탐 없이 처리됩니다. (2ee95fb)

**Comment by amondnet** [comment]
@cubic-dev-ai review

File: src/lib/asana-client.ts [hunk 4]
```diff
@@ -104,6 +104,52 @@ export function getAsanaClient() {
...
+        return result.data
+      },
+      setParent: async (taskGid: string, parentGid: string) => {
+        const body = { data: { parent: parentGid } }
+        const result = await tasksApiInstance!.setParentForTask(body, taskGid, {})
+        return result.data
+      },
+      // Dependency operations (this task is blocked by its dependencies)
+      getDependencies: async (taskGid: string, opts: any = {}) => {
+        const result = await tasksApiInstance!.getDependenciesForTask(taskGid, opts)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/lib/asana-client.ts]: [claude] @cubic-dev-ai 적용 완료 ✅ `getDependencies`에 `{ limit: 100, ...opts }` 기본 페이지네이션을 추가해 다른 task 래퍼 메서드(findAll/getSubtasks 등)와 일관되게 맞췄습니다. (63a9b24)

File: src/lib/asana-client.ts [hunk 5]
```diff
@@ -104,6 +104,52 @@ export function getAsanaClient() {
...
+        return result.data
+      },
+      removeDependencies: async (taskGid: string, dependencies: string[]) => {
+        const body = { data: { dependencies } }
+        const result = await tasksApiInstance!.removeDependenciesForTask(body, taskGid)
+        return result.data
+      },
+      // Dependent operations (these tasks are blocked by this task)
+      getDependents: async (taskGid: string, opts: any = {}) => {
+        const result = await tasksApiInstance!.getDependentsForTask(taskGid, opts)
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/lib/asana-client.ts]: [claude] @cubic-dev-ai 적용 완료 ✅ `getDependents`도 동일하게 `{ limit: 100, ...opts }`를 추가했습니다. 이로써 dependent 목록 누락과 합산 한도 검증 시 과소 카운트 위험이 해소됩니다. (63a9b24)

File: src/commands/task-dependency.ts [hunk 6]
```diff
@@ -0,0 +1,180 @@
+import chalk from 'chalk'
```
[the remark this answers is not in the document]
Comment (amondnet) [reply src/commands/task-dependency.ts:1]: [claude] @cubic-dev-ai 이번 PR에서는 스킵합니다 ⏭️ 빈 목록을 plain text(`No ... found`)로 출력하는 동작은 신규 명령만의 문제가 아니라 기존 `task list`/`section list`/`project list` 등 모든 list 명령의 공통 컨벤션입니다. 신규 명령만 바꾸면 오히려 비일관해지므로, CLI 전반의 빈 결과를 format-aware하게 만드는 별도 개선으로 추적하는 것이 적절합니다. 후속 이슈로 분리하겠습니다.
